Library/Homebrew/keg.rb

```diff
@@ -101,6 +102,19 @@ def initialize(path)
     @name = path.parent.basename.to_s
     @linked_keg_record = HOMEBREW_LIBRARY.join("LinkedKegs", name)
     @opt_record = HOMEBREW_PREFIX.join("opt", name)
+    begin
+      tab = Tab.from_file(path/"INSTALL_RECEIPT.json")
+      tap = tab.tap
+
+      oldname = if tap.nil? || tap == "Homebrew/homebrew"
+        Formulary.factory(name)
+      else
+        Formulary.factory("#{tap.sub("homebrew-", "")}/#{name}")
+      end.oldname
+
+      @oldname_opt_record = HOMEBREW_PREFIX.join("opt", oldname) if oldname
+    rescue Exception
+    end
```

Comment:
@xu-cheng I don't completely understand your idea:

* To get `oldname_opt_record` we need to:
```ruby
(opt_dir = HOMEBREW_PREFIX/"opt").directory?
  opt_dir.subdirs.detect do |dir|
    dir.symlink? && dir.resolved_path.parent == path.parent
  end
end
```
 There can be several `dir`s, so we have to exclude `opt_record` from `opt_dir.subdirs` before detecting.
* Before we can do the step above, the formula have to be migrated, so `oldname_opt_record` isn't `nil`.
Is this what you mean?